### ISSUE TITLE
[go_router] Clarification on Route Ordering with Multiple Matches: Follow-Up to Issue #106851

### ISSUE THREAD

**Issue by cizmarf** [flutter/flutter#150496]
### What package does this bug report belong to?

go_router

### What target platforms are you seeing this bug on?

Android, iOS, Web, macOS, Linux, Windows

### Have you already upgraded your packages?

Yes

### Dependency versions

<details><summary>pubspec.lock</summary>

```lock
go_router: ^14.1.2
```

</details>

### Steps to reproduce

In Issue #106851 (https://github.com/flutter/flutter/issues/106851), there was a promise of a documentation update about the importance of route order. Despite this, I cannot find any new information on the topic.

### Expected results

Please consider extending the documentation to include one of the following clarifications:
1. The most specific route is matched.
2. The first route in the list is matched.
3. It remains undocumented, but currently, the first route in the list is being matched. As this is an undocumented feature, it should not be relied upon, and GoRouter should warn users about this behavior.

### Actual results

No information found.

### Code sample

<details open><summary>Code sample</summary>

Practically I am not sure if it is safe to use
```dart
        routes: [
          GoRoute(
            path: 'specific_path',
            builder: (context, state) {
              return Page1();
            },
          ),
          GoRoute(
            path: ':id',
            builder: (context, state) {
              return Page2();
            },
          ),
```
where I `path: ':id',` can also match `path: 'specific_path',`.

</details>

### Screenshots or Videos

<details open>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details open><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.22.0, on macOS 14.3 23D5033f darwin-arm64
    (Rosetta), locale en-US)
    • Flutter version 3.22.0 on channel stable at
      /Users/filipcizmar/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 5dcb86f68f (6 weeks ago), 2024-05-09 07:39:20
      -0500
    • Engine revision f6344b75dc
    • Dart version 3.4.0
    • DevTools version 2.34.3

[✓] Android toolchain - develop for Android devices (Android SDK version
    34.0.0)
    • Android SDK at /Users/filipcizmar/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = /Users/filipcizmar/Library/Android/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      17.0.9+0-17.0.9b1087.7-11185874)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.4)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15F31d
    • CocoaPods version 1.14.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google
      Chrome

[✓] Android Studio (version 2023.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      17.0.9+0-17.0.9b1087.7-11185874)

[✓] VS Code (version 1.90.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.90.0

[✓] Connected device (5 available)            
    • Filip‘s iPhone (mobile)         • 00008130-000160A43041401C • ios
      • iOS 17.5.1 21F90
    • Filip‘s iPad (mobile)           • 00008103-0001684E0284C01E • ios
      • iOS 18.0 22A5282m
    • macOS (desktop)                 • macos                     •
      darwin-arm64   • macOS 14.3 23D5033f darwin-arm64 (Rosetta)
    • Mac Designed for iPad (desktop) • mac-designed-for-ipad     •
      darwin         • macOS 14.3 23D5033f darwin-arm64 (Rosetta)
    • Chrome (web)                    • chrome                    •
      web-javascript • Google Chrome 126.0.6478.62

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

**Comment by darshankawar** [comment]
Thanks for the report @cizmarf 
Have you already checked the readMe and migration guide to see if it is already covered ?

https://pub.dev/packages/go_router#documentation
https://pub.dev/packages/go_router#migration-guides

https://github.com/flutter/packages/blob/main/packages/go_router/CHANGELOG.md#1300

**Comment by cizmarf** [comment]
The only related mention I found is in 5.0 migration guide. But this seems to be focused on redirects, my issue is about `builder` or `pageBuilder`.

https://docs.google.com/document/d/10l22o4ml4Ss83UyzqUC8_xYOv_QjZEi80lJDNE4q7wM/edit?pli=1&resourcekey=0-U-BXBQzNfkk4v241Ow-vZg

> We also updated the redirection order when there are multiple redirections in route matches.
>
> In the previous version, the [GoRouter.redirect](https://pub.dev/documentation/go_router/latest/go_router/GoRouter-class.html) was called first. If there were multiple [GoRoute.redirect](https://pub.dev/documentation/go_router/latest/go_router/GoRoute/redirect.html)[s] in route matches. Only the last GoRoute.redirect was called while intermediate ones were ignored.
>
> After go_router 5.0,  the [GoRouter.redirect](https://pub.dev/documentation/go_router/latest/go_router/GoRouter-class.html) is still called first, but the [GoRoute.redirect](https://pub.dev/documentation/go_router/latest/go_router/GoRoute/redirect.html)[s] are called starting from the top-most GoRoute to the last GoRoute. The parent GoRoute.redirect takes priority over children’s. There is no migration plan for this change, and go_router won’t support the old behavior moving forward. If you need the old behavior, feel free to file an issue in [here](https://github.com/flutter/flutter/issues/new/choose) to discussion.

**Comment by darshankawar** [comment]
Thanks for the feedback. Please check if [this](https://github.com/flutter/flutter/issues/147716) resembles your case or not.

**Comment by cizmarf** [comment]
No, the issue https://github.com/flutter/flutter/issues/147716 covers a different case.

**Comment by darshankawar** [comment]
Thanks for the update. Keeping the issue open and labeling for team's attention.

**Comment by saurabh-mirajkar** [comment]
The team label got removed by `flutter-triage-bot`, so adding it again as per the triage guidelines.
